R/sampling.R

```diff
@@ -140,3 +140,121 @@ do_sampling <- function(data, p_control = 0.01) {
 
   dataall
 }
+
+
+#' @rdname sample_expanded_data
+setMethod(
+  f = "sample_expanded_data",
+  signature = "te_datastore",
+  definition = function(object, period, subset_condition, p_control, seed) {
+    data <- read_expanded_data(object, period = period, subset_condition = subset_condition)
+    data <- lapply(
+      split(data, list(data$trial_period, data$followup_time), drop = TRUE),
+      do_sampling,
+      p_control = p_control
+    )
+    data_table <- data.table::rbindlist(data)
+    data_table
+  }
+)
+
+
+#' @rdname sample_expanded_data
+setMethod(
+  f = "sample_expanded_data",
+  signature = "te_datastore_duckdb",
+  definition = function(object, period, subset_condition, p_control, seed) {
+    if (use_subset <- !missing(subset_condition)) {
+      subset_expr <- translate_to_sql(subset_condition)
+    }
+    q_p1 <- "SELECT * FROM (SELECT * FROM trial_data WHERE outcome = 0 "
+    q_p2 <- "UNION SELECT * FROM trial_data WHERE outcome = 1 "
+    if (is.null(seed)) {
+      q_sample <- paste0("USING SAMPLE ", p_control * 100, " PERCENT (bernoulli) ")
+    } else {
+      q_sample <- paste0("USING SAMPLE ", p_control * 100, " PERCENT (bernoulli, ", seed, ") ")
+    }
+    q_period <- ""
+    if (!is.null(period)) q_period <- paste0("AND trial_period IN (", paste0(period, collapse = ", "), ") ")
+    q_subset <- ""
+    if (use_subset) q_subset <- paste0("AND (", subset_expr, ")")
+
+    query <- paste0(q_p1, q_period, q_subset, ") ", q_sample, q_p2, q_period, q_subset)
+
+    data <- DBI::dbGetQuery(conn = object@con, statement = query)
+    data["sample_weight"] <- ifelse(data$outcome == 1, 1, 1 / p_control)
+    data_table <- data.table::as.data.table(data)
+    data_table
+  }
+)
+
+
+#' Translate subset_condition to SQL syntax
+#'
+#' @param string subset_condition as a string
+#'
+#' @return a string
+#'
+#' @noRd
+translate_to_sql <- function(string) {
+  replacement <- c("\\|" = "OR", "&" = "AND", "==" = "=", "%in%" = "IN", "^c\\(" = "\\(")
+  rt <- data.table::as.data.table(replacement, keep.rownames = "pattern")
+
+  vec <- strsplit(string, " ")[[1]]
+
+  if (length(grep(":", vec)) != 0) {
```

Comment:
```suggestion
  if (grepl(grep(":", vec))) {
```

Reply:
This throws an error, "argument "x" is missing, with no default"

Reply:
oops, yes I was not carelful. It needs to be
```suggestion
  if (grepl(":", vec)) {
```

Reply:
This throws an error too since multiple elements are returned (one TRUE/FALSE per element of the vector)

Reply:
ah, I see. Keep it as it is then